# Progetto Foundations of Computer Science
## Anno accademico 2021-22
### Elisa Bernasconi 815551, Damiano Garofoli 837589, Susanna Maugeri 839365

You have to work on the files:
*  [Books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Books.csv.gz)
*  [Book ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Book-Ratings.csv.gz)
*  [Users](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Users.csv.gz)
*  [Goodbooks books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks.csv.gz)
*  [Goodbooks ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks-ratings.csv.gz)

### Notes

1.    It is mandatory to use GitHub for developing the project.
1.    The project must be a jupyter notebook.
1.    There is no restriction on the libraries that can be used, nor on the Python version.
1.    To read those files, you need to use the `encoding = 'latin-1'` option.
1.    All questions on the project **must** be asked in a public channel on [Zulip](https://focs.zulipchat.com), otherwise no  answer will be given.

### Importazione delle librerie _pandas_ e _numpy_ e dei dataset necessari

In [1]:
import pandas as pd
import numpy as np

Iniziamo con l'importazione dei file .csv.

In [2]:
books = pd.read_csv("Data/Books.csv", encoding = 'latin-1', sep=';', low_memory=False, header=0)
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [3]:
book_rat = pd.read_csv("Data/Book-Ratings.csv", encoding = 'latin-1', sep=';', low_memory=False, header=0)
book_rat.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [4]:
users = pd.read_csv("Data/Users.csv", encoding = 'latin-1', sep=';', low_memory=False, header=0)
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
good_book = pd.read_csv("Data/goodbooks.csv", sep=',', encoding = 'latin-1', low_memory=False, header=0)
good_book.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   
1        2                  3             3  4640799          491  439554934   
2        3              41865         41865  3212258          226  316015849   
3        4               2657          2657  3275794          487   61120081   
4        5               4671          4671   245494         1356  743273567   

         isbn13                       authors  original_publication_year  \
0  9.780439e+12               Suzanne Collins                     2008.0   
1  9.780440e+12  J.K. Rowling, Mary GrandPré                     1997.0   
2  9.780316e+12               Stephenie Meyer                     2005.0   
3  9.780061e+12                    Harper Lee                     1960.0   
4  9.780743e+12           F. Scott Fitzgerald                     1925.0   

                             original_title  ... ratings_count  \
0                          The Hunger Games  ...       4780653   
1  Harry Potter and the Philosopher's Stone  ...       4602479   
2                                  Twilight  ...       3866839   
3                     To Kill a Mockingbird  ...       3198671   
4                          The Great Gatsby  ...       2683664   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1            4800065                    75867      75504     101676   
2            3916824                    95009     456191     436802   
3            3340896                    72586      60427     117415   
4            2773745                    51992      86236     197621   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1     455024    1156318    3011543   
2     793319     875073    1355439   
3     446835    1001952    1714267   
4     606158     936012     947718   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1474154022m...   
2  https://images.gr-assets.com/books/1361039443m...   
3  https://images.gr-assets.com/books/1361975680m...   
4  https://images.gr-assets.com/books/1490528560m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1474154022s...  
2  https://images.gr-assets.com/books/1361039443s...  
3  https://images.gr-assets.com/books/1361975680s...  
4  https://images.gr-assets.com/books/1490528560s...  

[5 rows x 23 columns]

In [6]:
good_rat = pd.read_csv("Data/goodbooks-ratings.csv", encoding = 'latin-1', sep=',', low_memory=False, header=0)
good_rat.head()

user_id  book_id  rating
0        1      258       5
1        2     4081       4
2        2      260       5
3        2     9296       5
4        2     2318       3

L'importazione dei file .csv è andata a buon fine, tuttavia nonostante l'utilizzo il tipo di encoding _Latin-1_ notiamo che vi sono dei caratteri che non sono stati decodificati nel modo corretto.

### 1. Normalize the location field of *Users* dataset, splitting into city, region, country.

Verifichiamo di che tipo sia ogni valore:

In [7]:
type(users['Location'][0])

str

Si tratta di stringhe, perciò per dividere i campi si può utilizzare uno split.

Verifichiamo che per tutte le righe vi siano le tre informazioni necessarie:

In [8]:
cont = np.zeros(10, dtype=int)

for i, row in users.iterrows():
    lung = len(row['Location'].split(','))
    if lung == 1:
        cont[0] += 1
    elif lung == 2:
        cont[1] += 1
    elif lung == 3:
        cont[2] += 1
    elif lung == 4:
        cont[3] += 1
    elif lung == 5:
        cont[4] += 1
    elif lung == 6:
        cont[5] += 1
    elif lung == 7:
        cont[6] += 1
    elif lung == 8:
        cont[7] += 1
    elif lung == 9:
        cont[8] += 1
    else:
        cont[9] += 1
        
for i in range(0,10):
    if i == 0:
        print('Righe con', i+1, 'elemento in Location:', cont[i])
    elif i == 9:
        print('Righe con 0, 10 o più elementi in Location:', cont[i])
    else:
        print('Righe con', i+1, 'elementi in Location:', cont[i])

Righe con 1 elemento in Location: 1
Righe con 2 elementi in Location: 1
Righe con 3 elementi in Location: 277348
Righe con 4 elementi in Location: 1417
Righe con 5 elementi in Location: 72
Righe con 6 elementi in Location: 11
Righe con 7 elementi in Location: 4
Righe con 8 elementi in Location: 2
Righe con 9 elementi in Location: 2
Righe con 0, 10 o più elementi in Location: 0


Il totale delle righe che non hanno l'esatto numero di elementi che ci servono sono

In [9]:
sbagliati = sum(cont)-cont[2]
sbagliati

1510

Il numero totale di righe della tabella _users_ è

In [10]:
totale = users.shape[0]
totale

278858

Perciò la percentuale di righe con informazioni mancanti o inesatte è del

In [11]:
print(round(sbagliati/totale*100, 2),'%')

0.54 %


Osserviamo nello specifico alcune di queste osservazioni:

In [12]:
print(users.loc[6117]['Location'])
print(users.loc[12512]['Location'])
print(users.loc[6457]['Location'])
print(users.loc[3924]['Location'])

birmingham, alabama, alabama, usa
nanaimo, bc, v9v 1h4, british columbia, canada
mayfield, sussex (haha, you listed stroud!), united kingdom
bombay, maharashtra, heard of it? :o), india


Alcune presentano valori ripetuti, senza significato o addirittura commenti personali.

Decidiamo pertanto di eliminare tali osservazioni dal dataset, poiché ne rappresentano una piccola percentuale e poiché potrebbero essere portatrici di informazioni errate o superflue. 

In [13]:
indici = []
for i, row in users.iterrows():
    lung = len(row['Location'].split(','))
    if lung != 3:
        indici.append(i)
        
users = users.drop(indici, axis=0)
len(users)

277348

Dunque il numero di osservazioni che consideriamo è 277348 e con queste possiamo ricavare le tre colonne richieste: _city, region_ e _country_.

In [14]:
users['city'] = users['Location'].str.split(expand=True, pat=',')[0]
users['region'] = users['Location'].str.split(expand=True, pat=',')[1]
users['country'] = users['Location'].str.split(expand=True, pat=',')[2]

In [15]:
users.head()

User-ID                            Location   Age         city  \
0        1                  nyc, new york, usa   NaN          nyc   
1        2           stockton, california, usa  18.0     stockton   
2        3     moscow, yukon territory, russia   NaN       moscow   
3        4           porto, v.n.gaia, portugal  17.0        porto   
4        5  farnborough, hants, united kingdom   NaN  farnborough   

             region          country  
0          new york              usa  
1        california              usa  
2   yukon territory           russia  
3          v.n.gaia         portugal  
4             hants   united kingdom

L'operazione è andata a buon fine.

### 2. For each book in the *Books* dataset, compute its average rating.

In [16]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [17]:
book_rat.columns

Index(['User-ID', 'ISBN', 'Book-Rating'], dtype='object')

Calcoliamo l'_average rating_ per ciascun libro della tabella _book_rat_ e poi operiamo un merge del risultato con la tabella _books_.

In [18]:
ave_rat = book_rat.groupby(['ISBN'])['Book-Rating'].mean()
ave_rat.head()

ISBN
 0330299891    3.0
 0375404120    1.5
 0586045007    0.0
 9022906116    3.5
 9032803328    0.0
Name: Book-Rating, dtype: float64

In [19]:
books = pd.merge(books, ave_rat, on = 'ISBN')
books.rename({'Book-Rating': 'Average-Rating'}, axis = 'columns', inplace = True)
books[['ISBN', 'Book-Title', 'Average-Rating']].head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

   Average-Rating  
0        0.000000  
1        4.928571  
2        5.000000  
3        4.272727  
4        0.000000

### 3. For each book in the *GoodBooks* dataset, compute its average rating.

Osserviamo i nomi di tutte le colonne del dataframe in questione:

In [20]:
good_book.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url'],
      dtype='object')

Notiamo che vi è già una colonna _average_rating,_ tuttavia proviamo a calcolare la media per verificare se i valori coincidono.

Consideriamo ad esempio gli elementi della riga con indice 0, sommiamo i valori delle colonne _ratings_1, ratings_2, ratings_3, ratings_4_ e _ratings_5_ per capire se come denominatore per la media debba essere utilizzato l'attributo _ratings_count_ oppure _work_ratings_count_.

In [21]:
print('Somma dei ratings: ', good_book['ratings_1'][0] + good_book['ratings_2'][0] + good_book['ratings_3'][0] + good_book['ratings_4'][0] + good_book['ratings_5'][0]) 
print('Ratings count: ', good_book['ratings_count'][0])
print('Work ratings count: ', good_book['work_ratings_count'][0])

Somma dei ratings:  4942365
Ratings count:  4780653
Work ratings count:  4942365


Capiamo che la somma delle 5 colonne di ratings è contenuta in _work_ratings_count_.

Calcoliamo dunque la media:

In [22]:
good_book['media'] = (good_book['ratings_1'] + good_book['ratings_2']*2 + good_book['ratings_3']*3 + good_book['ratings_4']*4 +good_book['ratings_5']*5)/good_book['work_ratings_count']
good_book['media'] = round(good_book['media'], ndigits=2)

In [23]:
good_book[['isbn', 'original_title', 'media', 'average_rating']]

isbn                            original_title  media  \
0    439023483                          The Hunger Games   4.34   
1    439554934  Harry Potter and the Philosopher's Stone   4.44   
2    316015849                                  Twilight   3.57   
3     61120081                     To Kill a Mockingbird   4.25   
4    743273567                          The Great Gatsby   3.89   
..         ...                                       ...    ...   
94   375751513                The Picture of Dorian Gray   4.06   
95   345803507                        Fifty Shades Freed   3.88   
96   393970124                                   Dracula   3.98   
97   307269981               Flickan som lekte med elden   4.22   
98  1612130585                       Fifty Shades Darker   3.87   

    average_rating  
0             4.34  
1             4.44  
2             3.57  
3             4.25  
4             3.89  
..             ...  
94            4.06  
95            3.88  
96            3.98  
97            4.22  
98            3.87  

[99 rows x 4 columns]

Effettivamente i valori del nostro calcolo, _media,_ e della colonna _average_rating_ coincidono.

### 4. Merge together all rows sharing the same book title, author and publisher. We will call the resulting datset `merged books`. The books that have not been merged together will not appear in `merged books`.

Con l'opzione _count_ otteniamo l'informazione circa quante osservazioni siamo state accorpate in ciascun gruppo di chiave _(Book-Title, Book-Author, Publisher)_.

In [24]:
merged_books = books.groupby(by=["Book-Title", "Book-Author", "Publisher"], as_index=False).count()
print(len(merged_books))
merged_books.head()

264961


Book-Title       Book-Author  \
0   A Light in the Storm: The Civil War Diary of ...       Karen Hesse   
1                              Always Have Popsicles    Rebecca Harvin   
2               Apple Magic (The Collector's series)  Martina Boudreau   
3   Ask Lily (Young Women of Faith: Lily Series, ...      Nancy N. Rue   
4   Beyond IBM: Leadership Marketing and Finance ...        Lou Mobley   

                     Publisher  ISBN  Year-Of-Publication  Image-URL-S  \
0  Hyperion Books for Children     1                    1            1   
1            Rebecca L. Harvin     1                    1            1   
2           Amer Cooking Guild     1                    1            1   
3                   Zonderkidz     1                    1            1   
4       Teleonet, Incorporated     1                    1            1   

   Image-URL-M  Image-URL-L  Average-Rating  
0            1            1               1  
1            1            1               1  
2            1            1               1  
3            1            1               1  
4            1            1               1

In [25]:
indici = merged_books[merged_books['ISBN'] == 1 ].index
merged_books.drop(indici, inplace=True)
print(len(merged_books))
merged_books.head()

4686


Book-Title  \
41                           !%@ (A Nutshell handbook)   
58   'A Hell of a Place to Lose a Cow': An American...   
252  10,000 dreams interpreted: A dictionary of dreams   
461                                   101 Famous Poems   
709                 15 Houseplants Even You Can't Kill   

                 Book-Author                       Publisher  ISBN  \
41             Donnalyn Frey                        O'Reilly     2   
58               Tim Brookes             National Geographic     2   
252  Gustavus Hindman Miller       Barnes &amp; Nobles Books     2   
461              Roy J. Cook  McGraw-Hill/Contemporary Books     3   
709                Joe Elder               Berkley Pub Group     2   

     Year-Of-Publication  Image-URL-S  Image-URL-M  Image-URL-L  \
41                     2            2            2            2   
58                     2            2            2            2   
252                    2            2            2            2   
461                    3            3            3            3   
709                    2            2            2            2   

     Average-Rating  
41                2  
58                2  
252               2  
461               3  
709               2

Così abbiamo ottenuto un dataframe contenente solo i 4686 libri che hanno avuto due o più edizioni.

Infine, eliminiamo le colonne _ISBN, Year-Of-Publication, Image-URL-S, Image-URL-M, Image-URL-L, Average-Rating,_ poiché esse non risultano essere più informative.

In [26]:
merged_books = merged_books.drop(['ISBN', 'Year-Of-Publication', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L', 'Average-Rating'], axis=1)
merged_books.head()

Book-Title  \
41                           !%@ (A Nutshell handbook)   
58   'A Hell of a Place to Lose a Cow': An American...   
252  10,000 dreams interpreted: A dictionary of dreams   
461                                   101 Famous Poems   
709                 15 Houseplants Even You Can't Kill   

                 Book-Author                       Publisher  
41             Donnalyn Frey                        O'Reilly  
58               Tim Brookes             National Geographic  
252  Gustavus Hindman Miller       Barnes &amp; Nobles Books  
461              Roy J. Cook  McGraw-Hill/Contemporary Books  
709                Joe Elder               Berkley Pub Group

### 5. For each book in `merged books` compute its average rating.

The average is computed considering all books in `books` that have been merged.

In [27]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L', 'Average-Rating'],
      dtype='object')

In [28]:
merged_books.columns

Index(['Book-Title', 'Book-Author', 'Publisher'], dtype='object')

Facciamo un _left join_ tra le tabelle _merged_books_ e _books,_ in modo tale che vengano tenuti tutti i libri che hanno avuto almeno due edizioni, con le informazioni relative a ciascuna edizione in righe diverse del dataframe.

In [29]:
merged_books2 = merged_books.merge(books, on=['Book-Title', 'Book-Author', 'Publisher'], how='left')
merged_books2.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1                          !%@ (A Nutshell handbook)            Donnalyn Frey   
2  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
3  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
4  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   

                   Publisher        ISBN Year-Of-Publication  \
0                   O'Reilly  1565920465                1994   
1                   O'Reilly  1565920317                1993   
2        National Geographic  0792276833                2000   
3        National Geographic  0792277295                2001   
4  Barnes &amp; Nobles Books  1566196256                1995   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/1565920465.0...   
1  http://images.amazon.com/images/P/1565920317.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792277295.0...   
4  http://images.amazon.com/images/P/1566196256.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/1565920465.0...   
1  http://images.amazon.com/images/P/1565920317.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792277295.0...   
4  http://images.amazon.com/images/P/1566196256.0...   

                                         Image-URL-L  Average-Rating  
0  http://images.amazon.com/images/P/1565920465.0...            0.00  
1  http://images.amazon.com/images/P/1565920317.0...            6.00  
2  http://images.amazon.com/images/P/0792276833.0...            0.00  
3  http://images.amazon.com/images/P/0792277295.0...            3.40  
4  http://images.amazon.com/images/P/1566196256.0...            7.25

Infine, possiamo calcolare l'_Average-Rating_ per le edizioni di ciascun libro.

In [30]:
ave_rat = merged_books2.groupby(['Book-Title', 'Book-Author', 'Publisher'])['Average-Rating'].mean()
merged_books = pd.merge(merged_books, ave_rat, on=['Book-Title', 'Book-Author', 'Publisher'])
merged_books.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
2  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   
3                                   101 Famous Poems              Roy J. Cook   
4                 15 Houseplants Even You Can't Kill                Joe Elder   

                        Publisher  Average-Rating  
0                        O'Reilly        3.000000  
1             National Geographic        1.700000  
2       Barnes &amp; Nobles Books        6.958333  
3  McGraw-Hill/Contemporary Books        3.111111  
4               Berkley Pub Group        0.000000

### 6. For each book in `merged books` compute the minimum and maximum of the average ratings over all corresponding books in the `books` dataset.

Hence for each book in `merged books` we will have exactly two values (a minimum and a maximum).

Utilizziamo la tabella _merged_books2_ che contiene le diverse edizioni di ciascun libro di _merged_books_ e creiamo dei nuovi dataframe, _minimi_ e _massimi,_ che contengono titolo, autore, editore di ciascun libro insieme al valore minimo o massimo dell'_average_rating._

In [31]:
minimi = merged_books2.groupby(['Book-Title', 'Book-Author', 'Publisher'], as_index=False)['Average-Rating'].min()
massimi = merged_books2.groupby(['Book-Title', 'Book-Author', 'Publisher'], as_index=False)['Average-Rating'].max()

minimi.rename({'Average-Rating': 'Minimum-Average-Rating'}, axis='columns', inplace=True)
massimi.rename({'Average-Rating': 'Maximum-Average-Rating'}, axis='columns', inplace=True)
massimi.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
2  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   
3                                   101 Famous Poems              Roy J. Cook   
4                 15 Houseplants Even You Can't Kill                Joe Elder   

                        Publisher  Maximum-Average-Rating  
0                        O'Reilly                    6.00  
1             National Geographic                    3.40  
2       Barnes &amp; Nobles Books                    7.25  
3  McGraw-Hill/Contemporary Books                    5.00  
4               Berkley Pub Group                    0.00

Creiamo dunque un unico dataframe che comprenda tutte le informazioni necessarie:

In [32]:
merged_books = pd.merge(merged_books, minimi, on=['Book-Title', 'Book-Author', 'Publisher'])
merged_books = pd.merge(merged_books, massimi, on=['Book-Title', 'Book-Author', 'Publisher'])
merged_books.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
2  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   
3                                   101 Famous Poems              Roy J. Cook   
4                 15 Houseplants Even You Can't Kill                Joe Elder   

                        Publisher  Average-Rating  Minimum-Average-Rating  \
0                        O'Reilly        3.000000                0.000000   
1             National Geographic        1.700000                0.000000   
2       Barnes &amp; Nobles Books        6.958333                6.666667   
3  McGraw-Hill/Contemporary Books        3.111111                0.000000   
4               Berkley Pub Group        0.000000                0.000000   

   Maximum-Average-Rating  
0                    6.00  
1                    3.40  
2                    7.25  
3                    5.00  
4                    0.00

### 7. For each book in `goodbooks`, compute the list of its authors. Assuming that the number of reviews with a text (column `work_text_reviews_count`) is split equally among all authors, find for each authors the total number of reviews with a text. We will call this quantity the *shared number of reviews with a text*.

Come prima cosa inizializziamo una nuova colonna per il dataframe _good_books_ che chiamiamo _List-Of-Authors_ e nella quale salviamo il risultato dello _split_ dei valori della colonna _authors._

In [33]:
good_book['List-Of-Authors'] = good_book['authors'].str.split(',')
good_book[['book_id', 'original_title', 'authors', 'List-Of-Authors']].head()

book_id                            original_title  \
0        1                          The Hunger Games   
1        2  Harry Potter and the Philosopher's Stone   
2        3                                  Twilight   
3        4                     To Kill a Mockingbird   
4        5                          The Great Gatsby   

                        authors                  List-Of-Authors  
0               Suzanne Collins                [Suzanne Collins]  
1  J.K. Rowling, Mary GrandPré  [J.K. Rowling,  Mary GrandPré]  
2               Stephenie Meyer                [Stephenie Meyer]  
3                    Harper Lee                     [Harper Lee]  
4           F. Scott Fitzgerald            [F. Scott Fitzgerald]

In [34]:
# !pip install gensim
from gensim.parsing.preprocessing import strip_multiple_whitespaces

Creiamo un dizionario in cui ad ogni autore è associato il numero di recensioni con testo che ha ricevuto, eventualmente divise per il numero di co-autori di ciascun libro.

In [35]:
autori = dict()
for ind, row in good_book.iterrows():
    for nome in row['List-Of-Authors']:
        nome = nome.strip()
        nome = strip_multiple_whitespaces(nome)
        if nome not in autori:
            autori[nome] = round(row['work_text_reviews_count']/len(row['List-Of-Authors']), 3)
        else:
            autori[nome] += row['work_text_reviews_count']/len(row['List-Of-Authors'])

In [36]:
shared_number_of_reviews_with_a_text_GB = autori
shared_number_of_reviews_with_a_text_GB

{'Suzanne Collins': 340066.0,
 'J.K. Rowling': 136668.0,
 'Mary GrandPré': 136668.0,
 'Stephenie Meyer': 258573.0,
 'Harper Lee': 72586.0,
 'F. Scott Fitzgerald': 51992.0,
 'John Green': 230584.0,
 'J.R.R. Tolkien': 52986.0,
 'J.D. Salinger': 44920.0,
 'Dan Brown': 66672.0,
 'Jane Austen': 52994.0,
 'Khaled Hosseini': 103375.0,
 'Veronica Roth': 156896.0,
 'George Orwell': 50644.667,
 'Erich Fromm': 15172.667,
 'Celâl Ã\x9cster': 15172.667,
 'Anne Frank': 6941.667,
 'Eleanor Roosevelt': 6941.667,
 'B.M. Mooyaart-Doubleday': 6941.667,
 'Stieg Larsson': 46983.0,
 'Reg Keeland': 46983.0,
 'Rufus Beck': 12033.0,
 'Alice Sebold': 36642.0,
 'William Golding': 26886.0,
 'William Shakespeare': 7389.0,
 'Robert Jackson': 7389.0,
 'Gillian Flynn': 121614.0,
 'Kathryn Stockett': 78204.0,
 'John Steinbeck': 24642.0,
 'Arthur Golden': 25605.0,
 'E.L. James': 128776.0,
 'Paulo Coelho': 27890.5,
 'Alan R. Clarke': 27890.5,
 'Lois Lowry': 54084.0,
 'C.S. Lewis': 15186.0,
 'Audrey Niffenegger': 43382

Per facilità di lettura lo trasformiamo in dataframe e rinominiamo le colonne in modo opportuno.

In [37]:
shared=pd.DataFrame(shared_number_of_reviews_with_a_text_GB.items())
shared.rename({0: "Author",1:"Number-Of-Reviews"},axis='columns',inplace=True)
shared

Author  Number-Of-Reviews
0      Suzanne Collins         340066.000
1         J.K. Rowling         136668.000
2       Mary GrandPré         136668.000
3      Stephenie Meyer         258573.000
4           Harper Lee          72586.000
..                 ...                ...
104        Oscar Wilde           9823.500
105  Jeffrey Eugenides           9823.500
106        Bram Stoker           5754.333
107      Nina Auerbach           5754.333
108      David J. Skal           5754.333

[109 rows x 2 columns]

### 8. For each year of publication, determine the author that has the largest value of the shared number of reviews with a text.

In [38]:
good_book.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url', 'media', 'List-Of-Authors'],
      dtype='object')

Creiamo un dizionario in cui le chiavi sono gli anni e i valori sono le liste di autori che hanno pubblicato uno o più libri in quell'anno.

In [39]:
anni_autori = dict()
for i, row in good_book.iterrows():
    anno = int(row['original_publication_year'])
    if anno not in anni_autori:
        anni_autori[anno] = row['authors'].strip().split(',')
        anni_autori[anno] = strip_multiple_whitespaces(row['authors'].strip()).split(',')
    else:
        anni_autori[anno].extend(row['authors'].strip().split(','))
        anni_autori[anno].extend(strip_multiple_whitespaces(row['authors'].strip()).split(','))
        
anni_autori

{2008: ['Suzanne Collins',
  'Stephenie Meyer',
  'Stephenie Meyer',
  'Stephenie Meyer',
  'Stephenie Meyer',
  'John Green',
  'John Green'],
 1997: ['J.K. Rowling', ' Mary GrandPré', 'Arthur Golden', 'Arthur Golden'],
 2005: ['Stephenie Meyer',
  'Stieg Larsson',
  ' Reg Keeland',
  'Stieg Larsson',
  ' Reg Keeland',
  'J.K. Rowling',
  ' Mary GrandPré',
  'J.K. Rowling',
  ' Mary GrandPré',
  'Rick Riordan',
  'Rick Riordan',
  'Markus Zusak',
  'Markus Zusak',
  'John Green',
  'John Green',
  'Jeannette Walls',
  'Jeannette Walls',
  'Steven D. Levitt',
  ' Stephen J. Dubner',
  'Steven D. Levitt',
  ' Stephen J. Dubner'],
 1960: ['Harper Lee'],
 1925: ['F. Scott Fitzgerald'],
 2012: ['John Green',
  'Gillian Flynn',
  'Gillian Flynn',
  'Veronica Roth',
  'Veronica Roth',
  'E.L. James',
  'E.L. James'],
 1937: ['J.R.R. Tolkien', 'John Steinbeck', 'John Steinbeck'],
 1951: ['J.D. Salinger'],
 2000: ['Dan Brown',
  'J.K. Rowling',
  ' Mary GrandPré',
  'J.K. Rowling',
  ' Mar

Da ogni lista eliminiamo i duplicati rendendola prima _set_ e poi nuovamente lista:

In [40]:
for key, value in anni_autori.items():
    anni_autori[key] = list(set(anni_autori[key]))
anni_autori

{2008: ['Suzanne Collins', 'Stephenie Meyer', 'John Green'],
 1997: ['J.K. Rowling', ' Mary GrandPré', 'Arthur Golden'],
 2005: ['J.K. Rowling',
  'Steven D. Levitt',
  'John Green',
  'Jeannette Walls',
  ' Stephen J. Dubner',
  ' Mary GrandPré',
  'Stieg Larsson',
  'Markus Zusak',
  'Stephenie Meyer',
  'Rick Riordan',
  ' Reg Keeland'],
 1960: ['Harper Lee'],
 1925: ['F. Scott Fitzgerald'],
 2012: ['Gillian Flynn', 'Veronica Roth', 'E.L. James', 'John Green'],
 1937: ['J.R.R. Tolkien', 'John Steinbeck'],
 1951: ['J.D. Salinger'],
 2000: ['J.K. Rowling', ' Mary GrandPré', 'Dan Brown'],
 1813: ['Jane Austen'],
 2003: ['J.K. Rowling',
  'Audrey Niffenegger',
  'Lauren Weisberger',
  'Khaled Hosseini',
  ' Mary GrandPré',
  'Mark Haddon',
  'Dan Brown'],
 2011: ['E.L. James', 'Veronica Roth'],
 1949: [' Erich Fromm', 'George Orwell', ' Celâl Ã\x9cster'],
 1945: ['George Orwell'],
 1947: [' Eleanor Roosevelt', ' B.M. Mooyaart-Doubleday', 'Anne Frank'],
 2009: ['Suzanne Collins', 'K

Osserviamo le chiavi del dizionario, che dovrebbero essere gli anni:

In [41]:
anni_autori.keys()

dict_keys([2008, 1997, 2005, 1960, 1925, 2012, 1937, 1951, 2000, 1813, 2003, 2011, 1949, 1945, 1947, 2009, 1999, 1954, 2010, 2002, 1998, 2007, 1595, 1988, 1993, 1950, 1996, 2006, 1868, 1847, 2001, 1953, 1974, 1979, 1932, 1884, 1952, 2015, 1995, 2004, 1969, 1936, 1985, 1818, 1977, 1811, -720, 1946, 1859, 1990, 1964, 1989, 1958, 1973, 1967, 1911, 1891, 1897])

Vi è un valore anomalo: -720. Osserviamo i titoli, gli autori e i codici ISBN a cui è associato questo anno di pubblicazione.

In [42]:
good_book[good_book['original_publication_year']== -720][['original_title', 'isbn', 'authors']]

original_title       isbn  \
78  Ὀδύσσεια  143039954   

                                              authors  
78  Homer, Robert Fagles, E.V. Rieu, Frédéric Mu...

Osserviamo che il primo autore risulta essere Omero e che il metodo di decoding _latin-1_ non ha correttamente decifrato il titolo di questo libro: ipotizziamo dunque che si possa trattare di un titolo in greco e che l'anno sia negativo in riferimento alla notazione a.C. e d.C. Come ulteriore verifica abbiamo cercato in rete il codice ISBN e abbiamo trovato che corrisponde ad una pubblicazione dell'Odissea tradotta da Fagles e introdotta da Knox ma pubblicata nel 1997. In ogni caso, dunque, l'anno di pubblicazione è inverosimile e non corretto e quindi decidiamo di eliminarlo dal dizionario.

In [43]:
del anni_autori[-720]
anni_autori.keys()

dict_keys([2008, 1997, 2005, 1960, 1925, 2012, 1937, 1951, 2000, 1813, 2003, 2011, 1949, 1945, 1947, 2009, 1999, 1954, 2010, 2002, 1998, 2007, 1595, 1988, 1993, 1950, 1996, 2006, 1868, 1847, 2001, 1953, 1974, 1979, 1932, 1884, 1952, 2015, 1995, 2004, 1969, 1936, 1985, 1818, 1977, 1811, 1946, 1859, 1990, 1964, 1989, 1958, 1973, 1967, 1911, 1891, 1897])

Ora non notiamo valori anomali.

Per ciascun anno mostriamo l'autore che ha il numero massimo di recensioni con testo in quell'anno e il relativo numero.

In [44]:
top_authors = []
for year in anni_autori.keys():
    libri_anno = good_book[good_book['original_publication_year'] == year] #dataframe
    autori_anno=dict()
    for i, riga in libri_anno.iterrows():
        autori_libro = riga['List-Of-Authors'] #lista di autori di ciascun libro
        for nome in autori_libro:
            nome = nome.strip() #pulizia del nome
            nome = strip_multiple_whitespaces(nome)
            if nome not in autori_anno:
                autori_anno[nome] = round(riga['work_text_reviews_count']/len(autori_libro), 3)
            else:
                autori_anno[nome] += riga['work_text_reviews_count']/len(autori_libro)
    valori = pd.DataFrame(autori_anno.items()) #coppie autore-numero di reviews
    indice = valori[1].idxmax()
    top_nome = valori.loc[indice][0]
    top_numero = valori.loc[indice][1]
    informazioni_top = [year, top_nome, top_numero]
    top_authors.append(informazioni_top)
    
Top_authors = pd.DataFrame(top_authors)
Top_authors.rename({0:'Year', 1:'Best-Author', 2:'Shared-Views-With-Text'}, axis='columns', inplace=True)
Top_authors.sort_values(by='Year', axis=0, ascending=True, inplace=True)
Top_authors.tail(15)

Year      Best-Author  Shared-Views-With-Text
16  1999  Stephen Chbosky                 47116.0
8   2000        Dan Brown                 25112.0
30  2001      Yann Martel                 42962.0
19  2002     Alice Sebold                 36642.0
10  2003  Khaled Hosseini                 59730.0
39  2004     Jodi Picoult                 30719.0
2   2005  Stephenie Meyer                 95009.0
27  2006       Sara Gruen                 55732.0
21  2007  Cassandra Clare                 51589.0
0   2008  Suzanne Collins                155254.0
15  2009  Suzanne Collins                 88538.0
18  2010  Suzanne Collins                 96274.0
11  2011       E.L. James                103489.0
5   2012       John Green                140739.0
37  2015    Paula Hawkins                 93600.0

### 9. Assuming that there are no errors in the ISBN fields, find the books in both datasets, and compute the difference of average rating according to the ratings and the goodratings datasets

Abbiamo già calcolato nei punti precedenti 2 e 3 l'_average_rating_ per i libri dei dataset _books_ e _good_book,_ perciò utilizziamo per questo calcolo le colonne delle tabelle che si chiamano, rispettivamente per _books_ e _good_book, Average-Rating_ e _average_rating._

In [45]:
intersection = pd.merge(books, good_book, left_on='ISBN', right_on='isbn') #how=inner è il default
intersection.rename({'Average-Rating': 'Average-Rating-B', 'average_rating': 'Average-Rating-GB'}, axis='columns', inplace=True)
intersection['Difference-Of-Ratings'] = intersection['Average-Rating-B'] - intersection['Average-Rating-GB']
intersection[['ISBN', 'Book-Title', 'Average-Rating-B', 'Average-Rating-GB', 'Difference-Of-Ratings']]

ISBN                                         Book-Title  \
0  1400032717  The Curious Incident of the Dog in the Night-T...   
1  014028009X                              Bridget Jones's Diary   
2  014038572X                      The Outsiders (Now in Speak!)   
3  1594480001                                    The Kite Runner   
4  043965548X  Harry Potter and the Prisoner of Azkaban (Harr...   

   Average-Rating-B  Average-Rating-GB  Difference-Of-Ratings  
0          4.813187               3.85               0.963187  
1          3.751852               3.75               0.001852  
2          4.460674               4.06               0.400674  
3          2.400000               4.26              -1.860000  
4          3.533333               4.53              -0.996667

### 10. Split the users dataset according to the age. One dataset contains the users with unknown age, one with age 0-14, one with age 15-24, one with age 25-34, and so on.

Controlliamo se vi sono valori _null_ per il campo _Age_ del dataset _users._

In [46]:
print(len(users))
len(users[users['Age'].isnull()])

277348


110204

Nel dataset _users,_ su 277348 osservazioni 110204 presentano un valore nullo _(NaN)_ per il campo _Age._ Definiamo un dataframe che contenga questi valori.

In [47]:
users_unknown = users[users['Age'].isnull() == True]
users_unknown.head()

User-ID                            Location  Age         city  \
0        1                  nyc, new york, usa  NaN          nyc   
2        3     moscow, yukon territory, russia  NaN       moscow   
4        5  farnborough, hants, united kingdom  NaN  farnborough   
6        7                 washington, dc, usa  NaN   washington   
7        8            timmins, ontario, canada  NaN      timmins   

             region          country  
0          new york              usa  
2   yukon territory           russia  
4             hants   united kingdom  
6                dc              usa  
7           ontario           canada

Definiamo una funzione che estrae dal dataset _users_ le osservazioni che hanno un valore di _Age_ compreso tra due valori dati.

In [48]:
def crea_dataset(a,b):
    dataset=users[(users['Age']>=a)&(users['Age']<=b)]
    return dataset

Per applicarla, verifichiamo quale sia il massimo valore assunto dalla variabile _Age_ nel dataset.

In [49]:
users['Age'].sort_values(ascending=False).head(15)

20856     244.0
8781      239.0
58285     237.0
1578      231.0
8457      230.0
172057    229.0
242213    228.0
192412    228.0
182769    228.0
163826    226.0
211520    223.0
245230    220.0
37834     219.0
52648     212.0
92857     210.0
Name: Age, dtype: float64

Ad un primo impatto si tratta di valori molto strani per quello che rappresentano, tuttavia non abbiamo la sicurezza che l'unità di misura utilizzata sia 'anni', anche se lo ipotizziamo. Verifichiamo quante osservazioni presentino valori più elevati di una simbolica età di 105 anni, per capire se essi possono essere considerati come valori anomali o _outliers_ che possono non essere considerati per questo task.

In [50]:
len(users[users['Age']>=106])

105

In [51]:
sopra_105=len(users[users['Age']>=106])
totali=len(users)

In [52]:
print(round(sopra_105/totali*100,3),'%')

0.038 %


Le osservazioni che presentano nel campo _Age_ un valore anomalo, maggiore della soglia simbolica dei 105 anni, sono 105. Poiché rappresentano una piccolissima parte del totale delle osservazioni, lo 0.04%, decidiamo di non considerare nella creazione di dataset per intervalli di 10 anni di età le osservazioni che presentano valori maggiori di 105.


Definiamo dunque i dataset richiesti:

In [53]:
users_0_14 = crea_dataset(0,14)
users_15_24 = crea_dataset(15,24)
users_25_34 = crea_dataset(25,34)
users_35_44 = crea_dataset(35,44)
users_45_54 = crea_dataset(45,54)
users_55_64 = crea_dataset(55,64)
users_65_74 = crea_dataset(65,74)
users_75_84 = crea_dataset(75,84)
users_85_94 = crea_dataset(85,94)
users_95_104 = crea_dataset(95,104)

E ne stampiamo uno di esempio:

In [54]:
users_35_44.head()

User-ID                         Location   Age        city  \
53       54            eubank, kentucky, usa  44.0      eubank   
66       67   framingham, massachusetts, usa  43.0  framingham   
69       70         rochester, new york, usa  44.0   rochester   
74       75      long beach, california, usa  37.0  long beach   
84       85  london, england, united kingdom  41.0      london   

            region          country  
53        kentucky              usa  
66   massachusetts              usa  
69        new york              usa  
74      california              usa  
84         england   united kingdom

### 11. Find the books that appear only in the goodbooks datasets.

Effettuiamo una _merge_ dei due dataset con le seguenti opzioni:
- _how = outer,_ che permette di considerare anche gli elementi che si trovano in uno solo dei due dataset che si uniscono;
- _indicator = True,_ che crea una variabile indicante in quale dei due dataset si trovasse l'osservazione _(left_only, right_only, both)._ Questa opzione ci permette successivamente di selezionare le osservazioni che si trovavano solo nel dataset _good_books,_ ovvero quello di destra.

In [55]:
union = pd.merge(books, good_book, left_on='ISBN', right_on='isbn', how='outer', indicator=True)
union[['ISBN','Book-Title','Book-Author','_merge']].head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author     _merge  
0    Mark P. O. Morford  left_only  
1  Richard Bruce Wright  left_only  
2          Carlo D'Este  left_only  
3      Gina Bari Kolata  left_only  
4       E. J. W. Barber  left_only

In [56]:
isbn_only_gb = union[union['_merge'] == 'right_only'][['isbn']]
isbn_only_gb.head()

isbn
270150  439023483
270151  439554934
270152  316015849
270153   61120081
270154  743273567

Una volta ottenuto un dataframe che contiene unicamente gli _isbn_ dei libri che si trovano solo in _good_book,_ effettuiamo un _inner join_ per estrarre e visualizzare anche il titolo e gli autori di ciascun libro.

In [57]:
books_only_gb = pd.merge(good_book, isbn_only_gb, on='isbn') #how='inner' è il default
books_only_gb[['isbn', 'original_title', 'authors']].head()

isbn                            original_title  \
0  439023483                          The Hunger Games   
1  439554934  Harry Potter and the Philosopher's Stone   
2  316015849                                  Twilight   
3   61120081                     To Kill a Mockingbird   
4  743273567                          The Great Gatsby   

                        authors  
0               Suzanne Collins  
1  J.K. Rowling, Mary GrandPré  
2               Stephenie Meyer  
3                    Harper Lee  
4           F. Scott Fitzgerald

### 12. Assuming that each pair (author, title) identifies a book, for each book find the number of times it appears in the books dataset. Which books appear the most times?

Per rispondere a questo quesito utilizziamo la funzione _groupby_ rispetto alla coppia _(Book-Author, Book-Title),_ conteggiamo la frequenza delle osservazioni e poi le ordiniamo per conteggio in ordine decrescente, per osservare il libro che appare più frequentemente.

In [58]:
grouped = books.groupby(['Book-Author', 'Book-Title'], as_index=False).count().sort_values(['ISBN'], ascending=False)
grouped.rename({'ISBN': 'Count'}, axis='columns', inplace=True)
grouped[['Book-Author', 'Book-Title', 'Count']].head()

Book-Author                      Book-Title  Count
149998  Louisa May Alcott                    Little Women     21
160597         Mark Twain  Adventures of Huckleberry Finn     20
69160        Emily Bronte               Wuthering Heights     20
106064        Jane Austen             Pride and Prejudice     18
36365     Charles Dickens              Great Expectations     16

Il libro che appare più frequentemente è _Little Women_ di Louisa May Alcott, per un totale di 21 volte.

### 13. Find the author with the highest average rating according to the goodbooks datasets.

Creiamo un dizionario nel quale al nome di ciascun autore sia associata una lista con i diversi ratings che ha ottenuto.

In [59]:
autori_ratings = dict()
for i, row in good_book.iterrows(): 
    for autore in row['List-Of-Authors']:
        autore = autore.strip()
        autore = strip_multiple_whitespaces(autore)
        if autore not in autori_ratings:
            autori_ratings[autore] = []
        autori_ratings[autore].append(row['average_rating'])
autori_ratings

{'Suzanne Collins': [4.34, 4.3, 4.03],
 'J.K. Rowling': [4.44, 4.53, 4.46, 4.37, 4.53, 4.61, 4.54],
 'Mary GrandPré': [4.44, 4.53, 4.46, 4.37, 4.53, 4.61, 4.54],
 'Stephenie Meyer': [3.57, 3.52, 3.69, 3.7, 3.84],
 'Harper Lee': [4.25],
 'F. Scott Fitzgerald': [3.89],
 'John Green': [4.26, 4.09, 3.88],
 'J.R.R. Tolkien': [4.25, 4.34],
 'J.D. Salinger': [3.79],
 'Dan Brown': [3.85, 3.79],
 'Jane Austen': [4.24, 4.06],
 'Khaled Hosseini': [4.26, 4.34],
 'Veronica Roth': [4.24, 4.07],
 'George Orwell': [4.14, 3.87],
 'Erich Fromm': [4.14],
 'Celâl Ã\x9cster': [4.14],
 'Anne Frank': [4.1],
 'Eleanor Roosevelt': [4.1],
 'B.M. Mooyaart-Doubleday': [4.1],
 'Stieg Larsson': [4.11, 4.22],
 'Reg Keeland': [4.11, 4.22],
 'Rufus Beck': [4.53],
 'Alice Sebold': [3.77],
 'William Golding': [3.64],
 'William Shakespeare': [3.73],
 'Robert Jackson': [3.73],
 'Gillian Flynn': [4.03],
 'Kathryn Stockett': [4.45],
 'John Steinbeck': [3.84],
 'Arthur Golden': [4.08],
 'E.L. James': [3.67, 3.88, 3.87],
 '

Successivamente, sostituiamo a ciascuna lista la media dei valori che contiene:

In [60]:
import statistics

for key, value in autori_ratings.items():
    media = statistics.mean(value)
    autori_ratings[key] = round(media, 2)

In [61]:
autori_ratings

{'Suzanne Collins': 4.22,
 'J.K. Rowling': 4.5,
 'Mary GrandPré': 4.5,
 'Stephenie Meyer': 3.66,
 'Harper Lee': 4.25,
 'F. Scott Fitzgerald': 3.89,
 'John Green': 4.08,
 'J.R.R. Tolkien': 4.29,
 'J.D. Salinger': 3.79,
 'Dan Brown': 3.82,
 'Jane Austen': 4.15,
 'Khaled Hosseini': 4.3,
 'Veronica Roth': 4.16,
 'George Orwell': 4.0,
 'Erich Fromm': 4.14,
 'Celâl Ã\x9cster': 4.14,
 'Anne Frank': 4.1,
 'Eleanor Roosevelt': 4.1,
 'B.M. Mooyaart-Doubleday': 4.1,
 'Stieg Larsson': 4.17,
 'Reg Keeland': 4.17,
 'Rufus Beck': 4.53,
 'Alice Sebold': 3.77,
 'William Golding': 3.64,
 'William Shakespeare': 3.73,
 'Robert Jackson': 3.73,
 'Gillian Flynn': 4.03,
 'Kathryn Stockett': 4.45,
 'John Steinbeck': 3.84,
 'Arthur Golden': 4.08,
 'E.L. James': 3.81,
 'Paulo Coelho': 3.82,
 'Alan R. Clarke': 3.82,
 'Lois Lowry': 4.12,
 'C.S. Lewis': 4.19,
 'Audrey Niffenegger': 3.95,
 'George R.R. Martin': 4.45,
 'Elizabeth Gilbert': 3.51,
 'Rick Riordan': 4.23,
 'Louisa May Alcott': 4.04,
 'Charlotte Brontë

Ed infine estraiamo l'autore che presenta il rating medio più alto.

In [62]:
autori_data_frame = pd.DataFrame(autori_ratings.items())
autori_data_frame.rename({0:'Author',1:'Average-Rating'}, axis='columns', inplace=True)
indice_max = autori_data_frame['Average-Rating'].idxmax()
autori_data_frame.loc[indice_max]['Author']

'Rufus Beck'

Con un _average_rating_ pari a:

In [63]:
autori_data_frame.loc[indice_max]['Average-Rating']

4.53